datafusion/src/datasource/csv.rs

```diff
@@ -46,10 +47,17 @@ use crate::physical_plan::csv::CsvExec;
 pub use crate::physical_plan::csv::CsvReadOptions;
 use crate::physical_plan::{common, ExecutionPlan};
 
+enum Source {
+    /// Path to a single CSV file or a directory containing one of more CSV files
+    Path(String),
+
+    /// Read CSV data from a reader
+    Reader(Mutex<Option<Box<dyn Read + Send + Sync + 'static>>>),
```

Comment:
How would this look if we would support multiple readers/partitions?

Reply:
Also I am wondering, if we do it like this, we would we need to do the same for json/xml/etc sources that should support essentially the same to avoid reimplementing it for each format.

Reply:
I suggest we file a follow on ticket for this work (supporting `Reader` input for JSON / XML sources). As you say I don't think it needs to be part of this PR.